File: codes/java/42.接雨水.java
class Solution {
	public int trap(int[] height) {
		if (height.length == 0) {
			return 0;
		}
		int n = height.length;
		int res = 0;
		// 数组充当备忘录
		int[] l_max = new int[n];
		int[] r_max = new int[n];
		// 初始化 base case
		l_max[0] = height[0];
		r_max[n - 1] = height[n - 1];
		// 从左向右计算 l_max
		for (int i = 1; i < n; i++)
			l_max[i] = Math.max(height[i], l_max[i - 1]);
		// 从右向左计算 r_max
		for (int i = n - 2; i >= 0; i--)
			r_max[i] = Math.max(height[i], r_max[i + 1]);
		// 计算答案
		for (int i = 1; i < n - 1; i++)
			res += Math.min(l_max[i], r_max[i]) - height[i];
		return res;
	}
}

File: codes/python/42.接雨水.py


class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        n = len(height)
        res = 0
        # 数组充当备忘录
        l_max = [0] * n
        r_max = [0] * n
        # 初始化 base case
        l_max[0] = height[0]
        r_max[n - 1] = height[n - 1]
        # 从左向右计算 l_max
        for i in range(1, n):
            l_max[i] = max(height[i], l_max[i - 1])
        # 从右向左计算 r_max
        for i in range(n - 2, -1, -1):
            r_max[i] = max(height[i], r_max[i + 1])
        # 计算答案
        for i in range(1, n - 1):
            res += min(l_max[i], r_max[i]) - height[i]
            """
            extend up -300
            ![](https://labuladong.github.io/pictures/接雨水/1.jpg)
            """
        return res